libutils/file_lib.c

```diff
@@ -650,6 +651,32 @@ int safe_chdir(const char *path)
 #endif // !__MINGW32__
 }
 
+int safe_chown_impl(const char *path, uid_t owner, gid_t group, int flags)
+{
+    int dirfd = -1;
+    int ret = -1;
+
+    char *parent_dir_alloc = xstrdup(path);
+    char *leaf_alloc = xstrdup(path);
+    char *parent_dir = dirname(parent_dir_alloc);
+    char *leaf = basename(leaf_alloc);
+
+    if ((dirfd = safe_open(parent_dir, O_RDONLY)) == -1)
+        goto bad;
+
+    if ((ret = fchownat(dirfd, leaf, owner, group, flags)))
+        goto bad;
```

Comment:
Code style (here and several other): We want braces around all if blocks, even one liners. I have fixed this for now, but remember it next time!